### ISSUE TITLE
No planes due to "MovingParam: ASSERT FAILED" with locale settings for comma as decimal point

### ISSUE THREAD

**Issue by kostex** [TwinFan/LiveTraffic#156]
**Describe the bug**
My installation has worked from the first release until a few weeks back. 
Log file shows:
LiveTraffic 1567459837.4 FATAL LTAircraft.cpp:133/MovingParam: ASSERT FAILED: defMax > defMin

**To Reproduce**
Message shows: Seeing xxx planes, displaying 0, etc..
Debugging shows correctly loading the CSL packages, but lots of assert failed

[Log.txt](https://github.com/TwinFan/LiveTraffic/files/3567133/Log.txt)

**Comment by TwinFan** [comment]
This is a support request better suited in the [Support Forum](https://forums.x-plane.org/index.php?/forums/forum/457-livetraffic-support/), not a bug report.

The assertion failing here is triggered on _creation_ of any new aircraft object. It will likely cause only few aircraft to appear, if any.

There are just two initializations of the `MovingParam` class, which aren't fed with constant values. (Constants would cause trouble with any user, me including, so we can rule them out.) The only non-constant usage takes values from the `FlightModels.prf` file. If that file is screwed up I could imagine the `assert` to trigger.

Can you make sure your `FlightModels.prf` file is as distributed in the [download package](https://forums.x-plane.org/index.php?/files/file/49749-livetraffic/). The file is located in the `Resources` folder.

There was no change to this part of the app in quite some while, so please verify your installation.

**Comment by kostex** [comment]
Sorry to have bothered you via this way.. found the Support forum.
I’ve checked every byte of the installation.. it’s a perfect match with the distributed package (except for the extra CSL planes of course)
To be certain OpenSky has some planes in the area (schiphol) I went to Opensky Explorer..
I did not change anything only to find out planes were appearing again..
The only thing I double checked is simulation time.. it was not on “track current time”
I guess that was the culprit?

Thanks for your efforts and support!

Regards,
Roel

**Comment by TwinFan** [comment]
Simulated time in XP is not supposed to have an effect. LiveTraffic works with actual time.

Is your issue solved?

**Comment by kostex** [comment]
Hi,
well.. I've tried on macOS and the planes appear perfectly fine..
looking at that DEBUG.log I get some "... reached invalid pos.. " but I
can see planes and it's all fine and dandy
on my Linux machine with the same settings/files I keep getting those
asserts and no planes are rendered.I hope it's not some kind of shared
library in Ubuntu that got updated or anything deep like that.. Debug
Log file is flooded with:
LiveTraffic 1567664390.9 FATAL LTAircraft.cpp:133/MovingParam: ASSERT
FAILED: defMax > defMinLiveTraffic 1567664390.9 FATAL
LTAircraft.cpp:133/MovingParam: ASSERT FAILED: defMax >
defMinLiveTraffic 1567664390.9 FATAL LTAircraft.cpp:133/MovingParam:
ASSERT FAILED: defMax > defMinLiveTraffic 1567664390.9 FATAL
LTAircraft.cpp:133/MovingParam: ASSERT FAILED: defMax >
defMinLiveTraffic 1567664390.9
WARN  LTFlightData.cpp:2147/CreateAircraft: Tracking data for '485814'
(man '', mdl '') lacks ICAO a/c type code, will be rendered with
standard a/c A320LiveTraffic 1567664390.9 FATAL
LTAircraft.cpp:133/MovingParam: ASSERT FAILED: defMax >
defMinLiveTraffic 1567664390.9
WARN  LTFlightData.cpp:2147/CreateAircraft: Tracking data for '485875'
(man '', mdl '') lacks ICAO a/c type code, will be rendered with
standard a/c A320LiveTraffic 1567664390.9 FATAL
LTAircraft.cpp:133/MovingParam: ASSERT FAILED: defMax >
defMinLiveTraffic 1567664390.9
WARN  LTFlightData.cpp:2147/CreateAircraft: Tracking data for '485A32'
(man '', mdl '') lacks ICAO a/c type code, will be rendered with
standard a/c A320LiveTraffic 1567664390.9 FATAL
LTAircraft.cpp:133/MovingParam: ASSERT FAILED: defMax >
defMinLiveTraffic 1567664390.9
WARN  LTFlightData.cpp:2147/CreateAircraft: Tracking data for '48C221'
(man '', mdl '') lacks ICAO a/c type code, will be rendered with
standard a/c A320LiveTraffic 1567664390.9 FATAL
LTAircraft.cpp:133/MovingParam: ASSERT FAILED: defMax >
defMinLiveTraffic 1567664390.9
WARN  LTFlightData.cpp:2147/CreateAircraft: Tracking data for '48C228'
(man '', mdl '') lacks ICAO a/c type code, will be rendered with
standard a/c A320LiveTraffic 1567664390.9 FATAL
LTAircraft.cpp:133/MovingParam: ASSERT FAILED: defMax >
defMinLiveTraffic 1567664390.9
WARN  LTFlightData.cpp:2147/CreateAircraft: Tracking data for '49D403'
(man '', mdl '') lacks ICAO a/c type code, will be rendered with
standard a/c A320LiveTraffic 1567664390.9 FATAL
LTAircraft.cpp:133/MovingParam: ASSERT FAILED: defMax >
defMinLiveTraffic 1567664390.9 FATAL LTAircraft.cpp:133/MovingParam:
ASSERT FAILED: defMax > defMinLiveTraffic 1567664390.9 FATAL
LTAircraft.cpp:133/MovingParam: ASSERT FAILED: defMax >
defMinLiveTraffic 1567664390.9 FATAL LTAircraft.cpp:133/MovingParam:
ASSERT FAILED: defMax > defMinLiveTraffic 1567664390.9 FATAL
LTAircraft.cpp:133/MovingParam: ASSERT FAILED: defMax >
defMinLiveTraffic 1567664390.9
WARN  LTFlightData.cpp:2147/CreateAircraft: Tracking data for '4CACB5'
(man '', mdl '') lacks ICAO a/c type code, will be rendered with
standard a/c A320LiveTraffic 1567664390.9 FATAL
LTAircraft.cpp:133/MovingParam: ASSERT FAILED: defMax >
defMinLiveTraffic 1567664390.9 FATAL LTAircraft.cpp:133/MovingParam:
ASSERT FAILED: defMax > defMinLiveTraffic 1567664390.9
WARN  LTFlightData.cpp:2147/CreateAircraft: Tracking data for '4CAFA6'
(man '', mdl '') lacks ICAO a/c type code, will be rendered with
standard a/c A320LiveTraffic 1567664390.9 FATAL
LTAircraft.cpp:133/MovingParam: ASSERT FAILED: defMax >
defMinLiveTraffic 1567664390.9
WARN  LTFlightData.cpp:2147/CreateAircraft: Tracking data for '4D21D4'
(man '', mdl '') lacks ICAO a/c type code, will be rendered with
standard a/c A320LiveTraffic 1567664390.9 FATAL
LTAircraft.cpp:133/MovingParam: ASSERT FAILED: defMax >
defMinLiveTraffic 1567664390.9 FATAL LTAircraft.cpp:133/MovingParam:
ASSERT FAILED: defMax > defMin

**Comment by TwinFan** [comment]
So it is now a Linux-only issue, but reproducible? I don’t have a Linux environment for testing. Build happens in a Docker environment. But as the issue affects one definite line of code I could add some diagnostics around and ask you to try that version. Would you support?

**Comment by kostex** [comment]
Sure! Would be glad to be of assistance!

Roel

Op do 5 sep. 2019 12:09 schreef TwinFan <notifications@github.com>:

**Comment by TwinFan** [comment]
Please try the attached version
[lin.xpl.zip](https://github.com/TwinFan/LiveTraffic/files/3585681/lin.xpl.zip)
with Log level set to Debug (otherwise output wouldn't show up in Log.txt) and upload the resulting `Log.txt` file.

A successful plane creation would look like this now:
```
LiveTraffic 1567801930.6 DEBUG LTFlightData.cpp:2152/CreateAircraft: About to create an a/c A9DCE9 of type B739
LiveTraffic 1567801930.6 DEBUG LTAircraft.cpp:806/FindFlightModel: a/c type B739, string to match: M;L2J;B739;737-900;BOEING
LiveTraffic 1567801930.6 DEBUG LTAircraft.cpp:813/FindFlightModel: Found flight model MediumJets
LiveTraffic 1567801930.6 DEBUG LTAircraft.cpp:134/MovingParam: _dur=10.00, _max=1.00, _min=0.00, _wrap=0
LiveTraffic 1567801930.6 DEBUG LTAircraft.cpp:134/MovingParam: _dur=5.00, _max=1.00, _min=0.00, _wrap=0
LiveTraffic 1567801930.6 DEBUG LTAircraft.cpp:134/MovingParam: _dur=45.00, _max=360.00, _min=0.00, _wrap=1
LiveTraffic 1567801930.6 DEBUG LTAircraft.cpp:134/MovingParam: _dur=12.00, _max=30.00, _min=-30.00, _wrap=0
LiveTraffic 1567801930.6 DEBUG LTAircraft.cpp:134/MovingParam: _dur=3.40, _max=15.00, _min=-2.00, _wrap=0
LiveTraffic 1567801930.6 DEBUG LTAircraft.cpp:134/MovingParam: _dur=2.00, _max=1.00, _min=0.00, _wrap=0
LiveTraffic 1567801930.6 DEBUG LTAircraft.cpp:134/MovingParam: _dur=0.50, _max=1.00, _min=0.00, _wrap=0
LiveTraffic 1567801930.6 DEBUG LTAircraft.cpp:134/MovingParam: _dur=5.00, _max=2000.00, _min=0.00, _wrap=0
LiveTraffic 1567801930.6 DEBUG LTAircraft.cpp:134/MovingParam: _dur=0.50, _max=0.50, _min=0.00, _wrap=0
LiveTraffic 1567801930.6 INFO  LTAircraft.cpp:954/LTAircraft: Added aircraft A9DCE9 (B739 UA1521), operator 'UAL', a/c model 'BB_Boeing B739_UAL', flight model [MediumJets], bearing 242, distance 19.1nm, from channel OpenSky Live Online
```

while an unsuccessful creation would look like this:

```
LiveTraffic 1567802128.3 DEBUG LTFlightData.cpp:2152/CreateAircraft: About to create an a/c A10597 of type DA40
LiveTraffic 1567802128.3 DEBUG LTAircraft.cpp:806/FindFlightModel: a/c type DA40, string to match: L;L1P;DA40;Club Star;DIAMOND
LiveTraffic 1567802128.3 DEBUG LTAircraft.cpp:813/FindFlightModel: Found flight model GA
LiveTraffic 1567802128.3 DEBUG LTAircraft.cpp:134/MovingParam: _dur=10.00, _max=1.00, _min=0.00, _wrap=0
LiveTraffic 1567802128.3 FATAL LTAircraft.cpp:135/MovingParam: ASSERT FAILED: defMax < defMin
LiveTraffic 1567802128.3 ERROR LTFlightData.cpp:2158/CreateAircraft: Exception occured while creating a/c A10597 of type DA40: LiveTraffic 1567802128.3 FATAL LTAircraft.cpp:135/MovingParam: ASSERT FAILED: defMax < defMin
LiveTraffic 1567802128.3 ERROR LTFlightData.cpp:2168/CreateAircraft: Could not create new object (memory?): A10597
```

Also the exception as such should be handled more gracefully, which is code which might even stay in.

**Comment by kostex** [comment]
Hi there,
here's the log.zip
I noticed it kept on flooding so the log is quite big.. and repeats
itself I guess.. but you'll figure it out I'm sure ;-)
Cheers,Roel

**Comment by TwinFan** [comment]
It will repeat over and over...but maybe you forgot to actually attach the log? ;)

**Comment by kostex** [comment]
I most certainly did not forget to attach it.. but I replied with some dubious mail application on a remote Linux machine.. and looking directly here at Github it's not attached.. hmm.. weird.. but here you go:
[Log.zip](https://github.com/TwinFan/LiveTraffic/files/3585953/Log.zip)

**Comment by TwinFan** [comment]
OK...I believe I've got it...you are running some European settings also for number formats, could that be?

Already from the very first Log.txt I was surprised to read
```
LiveTraffic: LiveTraffic 1,22.190906 starting up...
```
with a **comma** between 1 and 22 instead of a dot. All my logs and those of other users read `1.22.190906`. At first I thought you compiled yourself...but the date matched official release, so I left the thought there.

Now, however, we need to pick up that thought again: Aircraft creation fails when loading the `GEAR_DEFLECTION` parameter. It is the only parameter in the `FlightModel.prf` file which uses decimal places for good. It reads
```
GEAR_DEFLECTION        0.5  # [m] main gear deflection during touch-down
```
With your European settings, expecting a comma as decimal "point", this would probably be read as `0` instead of `0,5`. And then the assert fails: `0` (which was supposed to be `0.5`) is not greater than `0`.

The `GEAR_DEFLECTION` parameter is new with v1.22, which explains why the issue came up only now.

As a quick fix, here's a `FlightModels.prf` with all decimal places removed. Your planes might dip a bit deeper upon touch down, but otherwise operations should be fine:
[FlightModels.prf.zip](https://github.com/TwinFan/LiveTraffic/files/3585975/FlightModels.prf.zip)

Not sure how I fix that in the long run across all platforms with their thousands of different settings...I'd rather go and interpret the decimal places myself or change units so I don't need them altogether (ie. specify `cm` rather than `m`).

Also, I need to enhance validation: If you'd specify a `GEAR_DEFLECTION` of 0 (which would, for example, be appropriate for planes without gear (struts) like most sail planes) the assert would still fail the same way.

So, well...this ends up being a bug in the end...and GitHub is the right place to report it ;-) It would be a confirmed bug if you can confirm LiveTraffic working with the provided `FlightModels.prf`.

**Comment by kostex** [comment]
Pff... sounds familiar.. comma and dot issues.. They keep on haunting us since computers needed to be outfitted with locale settings..
And yes I'm in Europe.. Netherlands in fact.. started up the sim and thankfully in the middle of the night some planes were buzzing around at Schiphol ;-)
**And YES! They showed up in/with LiveTraffic!! Whoohoo!**

So your modified FlightModels.prf makes me a happy camper!

Before I reported all this I was really under the assumption that the only way to see planes was to get a subscription and therefore removed the addon.. but XPlane felt very empty so I picked it up again to try to figure out what was wrong.. good good!

Thanks a bunch for taking my question/report serious.. it get things fixed so we see!

/me bug reporting in the middle of the night while filled up with beer and wine.. great...
